examples/DGmethods/held_suarez_forcing.jl

```diff
@@ -0,0 +1,1017 @@
+    # F[3, _dρe] -= u * τ31 + v * τ32 + w * τ33 + ν_e * k_μ * vTz
+  end
+end
+#md nothing # hide
+
+# FXG: Source function => Define the geopotential and Held-Suarez source from the solution and auxiliary variables
+function source!(S, Q, aux, t)
+  @inbounds begin
+
+    ## Store values
+    dρ, ρu, ρv, ρw, dρe = Q[_dρ], Q[_ρu], Q[_ρv], Q[_ρw], Q[_dρe]
+    ρ_ref, ρe_ref = aux[_a_ρ_ref], aux[_a_ρe_ref]
+    ϕ, ϕx, ϕy, ϕz = aux[_a_ϕ], aux[_a_ϕx], aux[_a_ϕy], aux[_a_ϕz]
+    x, y, z = aux[_a_x], aux[_a_y], aux[_a_z]
+
+    ## Add Geopotential source
+    S[_dρ ] = 0
+    if PDE_level_hydrostatic_balance
+      S[_ρu ] = -dρ * ϕx
+      S[_ρv ] = -dρ * ϕy
+      S[_ρw ] = -dρ * ϕz
+    else
+      ρ = ρ_ref + dρ
+      S[_ρu ] = -ρ * ϕx
+      S[_ρv ] = -ρ * ϕy
+      S[_ρw ] = -ρ * ϕz
+    end
+    S[_dρe] = 0
+    
+    # Coriolis force
+    if use_coriolis
+      coriolis_x =  2Omega * ρv
+      coriolis_y = -2Omega * ρu
+      coriolis_z =  0
+
+      S[_ρu ] += coriolis_x
+      S[_ρv ] += coriolis_y
+      S[_ρw ] += coriolis_z
+    end
+
+    ## Add Held-Suarez Source
+    #Extract Temperature
+    ρ = ρ_ref + dρ
+    ρe = ρe_ref + dρe
+    e = ρe / ρ
+    u, v, w = ρu / ρ, ρv / ρ, ρw / ρ
+    e_int = e - (u^2 + v^2 + w^2)/2 - ϕ
+    T = air_temperature(e_int)
+    P = air_pressure(T, ρ)
+    
+    #Create Held-Suarez forcing
+    (kv,kt,T_eq)=held_suarez_forcing(x,y,z,T,P,t) 
+    
+    #Apply forcing
+    if use_held_suarez_forcings
+      S[_ρu ]  -= kv*ρu
+      S[_ρv ]  -= kv*ρv
+      S[_ρw ]  -= kv*ρw
+      S[_dρe ] -= ( kt*ρ*cv_d*T + kv*(ρu*ρu + ρv*ρv + ρw*ρw)/ρ )
+    end
+
+    if use_sponge
+      sponge_coefficient = aux[_a_sponge]
+      S[_ρu ] -= sponge_coefficient*ρu
+      S[_ρv ] -= sponge_coefficient*ρv
+      S[_ρw ] -= sponge_coefficient*ρw
+    end
+end
+end
+#md nothing # hide
+
+# FXG: Held-Suarez forcing function
+function held_suarez_forcing(x, y, z, T, P, t)
+  @inbounds begin
+    DFloat = eltype(T)
+
+    #Store Held-Suarez constants
+    p0 :: DFloat = MSLP
+    θ0 :: DFloat = 315
+    N_bv :: DFloat = 0.0158725
+    gravity :: DFloat = grav
+    ka :: DFloat = 1 / 40 / 86400
+    kf :: DFloat = 1 / 86400
+    ks :: DFloat = 1 / 4 / 86400
+    ΔT_y :: DFloat = 60 
+    ΔT_z :: DFloat = 10   
+    temp0 :: DFloat = 315
+    temp_min :: DFloat = 200
+    hd :: DFloat = 7000 #from Smolarkiewicz JAS 2001 paper    
+    σ_b :: DFloat = 7 / 10
+    
+    #Compute Rayleigh Damping terms from Held-Suarez 1994 paper
+    (r, λ, φ) = cartesian_to_spherical(x, y, z)
+    h = r - DFloat(planet_radius) # height above the planet surface
+    σ = exp(-h/hd) #both approx of sigma behave similarly
+#    σ = P/p0        #both approx of sigma behave similarly
+    Δσ = (σ - σ_b)/(1 - σ_b)
+    π = σ^(R_d/cp_d)
+    c = max(0, Δσ)
+    T_eq = ( temp0 - ΔT_y*sin(φ)^2 - ΔT_z*log(σ)*cos(φ)^2 )*π
+    T_eq = max(temp_min, T_eq)
+    kt = ka + (ks-ka)*c*cos(φ)^4
+    kv = kf*c
+    kt = kt*(1 - T_eq/T)
```

Comment:
It's quite confusing to have the part of the source that involves the state variable (or aux variable) here, and the part involving the reference state in another part of the code. Makes it difficult to piece together what is going on.